### PR TITLE
Add `WithSampler` option

### LINKED ISSUE #454: Add new option method `WithSampler`
**Is your feature request related to a problem? Please describe.**

We want to use the `opentelemetry-go-instrumentation` to auto-instrument our Go microservices. To reduce storage costs and provide a better sampling strategy, we want something that allows us to exclude some traces when their request path (e.g. `/health`) matches a regular expression.

This could be very useful in environments like Kubernetes or similar where probes are regularly sent to the app.

**Describe the solution you'd like**

My first idea was to write a custom Go-specific sampler, `RequestPathSampler`, that takes a simple list of regular expressions as a parameter. For instance, when I configure my auto instrumentation, I would like to have : 

```
export OTEL_TRACES_SAMPLER="request_path_sampler"
export OTEL_TRACES_SAMPLER_ARG="/(health|metrics)"
```

> I would prefer to keep if possible `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG` to ensure consistency with [OpenTelemetry built-in samplers configuration](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_traces_sampler).

When configured with the environment variables above, all traces matching the given pattern will be automatically dropped at the instrumentation level. 

I dug a little into the code and saw that in [`instrumentation.go`](https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/main/instrumentation.go#L219), we force the usage of the `trace.AlwaysSample` sampler. Maybe we can update a little this part of the code, in addition to adding the `RequestPathSampler` code to have something more modular for the sampler implementation, through a factory for instance.

I can make a PR for this proposal if needed.

**Describe alternatives you've considered**

Currently, I haven't found any viable workaround to drop some traces based on the path. I forked the repository into the company's GitLab to add this feature, but this is not ideal at all because I need to maintain my version of the project in addition to the feature. It would be better to have it included in the upstream repository.

**Additional context**

Related issues: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1060

### PR DESCRIPTION
**Pull request by thomasgouveia**
This PR adds a new built-in sampler to filter traces based on the HTTP request path. 

This feature could be handy when deploying into orchestration systems like **Kubernetes** that produce a trace for each liveness/readiness probe and could lead to higher storage costs.

Fixes #454

### CODE REVIEW

**Comment by thomasgouveia** [comment]
Hi @MrAlias, If I understand correctly, you want to : 

- Move the implementation of the `HttpFilterOut` sampler into https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main
- Rename the `WithPathFilter` method to `WithSampler`

LGTM in that case because it allows flexibility and customization for anyone who wants to build a custom auto instrumentation binary.

**Comment by MrAlias** [comment]
+1 to `WithSampler`.

I think the contrib addition is up for discussion. I think there are a few models for that addition. We would likely want to support environment variables there.

For this though, I think adding a sampler option is a good start. 👍

**Comment by thomasgouveia** [comment]
@MrAlias I've just updated the PR with what we discussed. I've also updated the related issue. Let me know if something is missing!

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 - Add the `WithTraceExporter` `InstrumentationOption` to configure the trace `SpanExporter` used by an `Instrumentation`. ([#426](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/426))
 - Add HTTP status code attribute to `net/http` server instrumentation. ([#428](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/428))
 - The instrumentation scope now includes the version of the auto-instrumentation project. ([#442](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/442))
+- Add a new `WithSampler` method allowing end-users to provide their own implementation of OpenTelemetry sampler directly through the package API. ([#454](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/468)).
```
[the remark this answers is not in the document]
Comment (thomasgouveia) [reply CHANGELOG.md]: 👍

**Review by MrAlias — APPROVED** [review]

**Review by pellared — APPROVED** [review]

**Pull request merged**